examples/camera-app/linux/src/clusters/camera-avsettingsuserlevel-mgmt/camera-avsettingsuserlevel-manager.cpp

```diff
@@ -92,10 +147,61 @@ Status CameraAVSettingsUserLevelManager::MPTZRemovePreset(uint8_t aPreset)
 Status CameraAVSettingsUserLevelManager::DPTZSetViewport(uint16_t aVideoStreamID, Structs::ViewportStruct::Type aViewport)
 {
     // The Cluster implementation has ensured that the videoStreamID represents a valid stream.
-    // The application needs to interact with its instance of AVStreamManagement to access the stream, validate the viewport
-    // and set the new vieport value.
+    // The application needs to interact with iHAL to access the stream, validate the viewport
+    // and set the new viewport value.
     //
-    return Status::Success;
+    Status status    = Status::Success;
+    bool streamFound = false;
+
+    for (VideoStream & stream : mCameraDeviceHAL->GetAvailableVideoStreams())
+    {
+        if (stream.videoStreamParams.videoStreamID == aVideoStreamID && stream.isAllocated)
+        {
+            streamFound = true;
+            // Validate the received viewport dimensions
+            //
+            // Ensure pixel count is > min pixels
+            // Esnure width does not exceed sensor width
+            // Ensure height does not exceed sensor height
+            //
+            uint16_t requestedWidth             = aViewport.x2 - aViewport.x1;
+            uint16_t requestedHeight            = aViewport.y2 - aViewport.y1;
+            VideoResolutionStruct minResolution = mCameraDeviceHAL->GetMinViewport();
+            VideoSensorParamsStruct sensorParms = mCameraDeviceHAL->GetVideoSensorParams();
+            if ((requestedWidth < minResolution.width) || (requestedHeight < minResolution.height) ||
+                (requestedWidth > sensorParms.sensorWidth) || (requestedHeight > sensorParms.sensorHeight))
+            {
+                ChipLogError(Camera, "CameraApp: DPTZSetViewport with invalid viewport dimensions");
+                status = Status::ConstraintError;
+                break;
+            }
+
+            // Get the ARs to no more than 2DP.  Otherwise you get mismatches e.g. 16:9 ratio calculation for 480p isn't the same as
+            // 1080p beyond 2DP.
+            float requestedAR = floorf((static_cast<float>(requestedWidth) / requestedHeight) * 100) / 100;
+            float streamAR    = floorf((static_cast<float>(stream.videoStreamParams.maxResolution.width) /
+                                     stream.videoStreamParams.maxResolution.height) *
+                                       100) /
+                100;
+
+            ChipLogDetail(Camera, "DPTZSetViewpoort. AR of viewport %f, AR of stream %f.", requestedAR, streamAR);
+            // Ensure that the aspect ration of the viewport matches the aspect ratio of the stream
+            if (requestedAR != streamAR)
+            {
+                ChipLogError(Camera, "CameraApp: DPTZSetViewport with mismatching aspect ratio.");
+                status = Status::ConstraintError;
+                break;
+            }
+            mCameraDeviceHAL->SetViewport(stream, aViewport);
+        }
+    }
+
+    if (!streamFound)
+    {
```

Comment:
if we use early return, then we can jusr
   
   return Status::InvalidCommand;   // stream not found